New features

OlivierLDff opened the pull request:
This fork aim is to add new functionalities.

* Customizable **VERSION_NAME** that can be set during the macro call.
* Allow to deploy a mix between having a **PACKAGE_SOURCES** and using a configurable `AndroidManifest.xml.in`.
* Specify a custom `AndroidManifest.xml.in` with **ANDROID_MANIFEST_IN_PATH** during the macro call.
* Warning when a static or none **ANDROID_STL** is specified.
* Move the *minSdkVersion* and *targetSdkVersion* from `AndroidManifest.xml.in` to a `build.gradle.in` to match new Android Policy.
* Update the `AndroidManifest.xml.in` to the one in `Qt5.12.x`. `qtandroiddeploy` will add the required permission depending on what is deployed.
* The apk is build inside a folder `${CMAKE_CURRENT_BINARY_DIR}/${SOURCE_TARGET}-${ANDROID_ABI}`.
* Specify the `--release` or `--debug` flag to `qtandroiddeploy`.
  - `--release` : *Release | MinSizeRel | RelWithDebInfo*
  - `--debug` : *Debug*
* `--verbose` can now be specified with the **VERBOSE** macro argument.

For any question about this fork feel free to [contact me](olivier.ldff@gmail.com).

Tested with `qt5.11 qt5.12` on `windows 1803`.

OlivierLDff committed:
Extend with new functionalities

This fork aim is to add new functionalities.

* Customizable **VERSION_NAME** that can be set during the macro call.
* Allow to deploy a mix between having a **PACKAGE_SOURCES** and using a configurable `AndroidManifest.xml.in`.
* Specify a custom `AndroidManifest.xml.in` with **ANDROID_MANIFEST_IN_PATH** during the macro call.
* Warning when a static or none **ANDROID_STL** is specified.
* Move the *minSdkVersion* and *targetSdkVersion* from `AndroidManifest.xml.in` to a `build.gradle.in` to match new Android Policy.
* Update the `AndroidManifest.xml.in` to the one in `Qt5.12.x`. `qtandroiddeploy` will add the required permission depending on what is deployed.
* The apk is build inside a folder `${CMAKE_CURRENT_BINARY_DIR}/${SOURCE_TARGET}-${ANDROID_ABI}`.
* Specify the `--release` or `--debug` flag to `qtandroiddeploy`.
  - `--release` : *Release | MinSizeRel | RelWithDebInfo*
  - `--debug` : *Debug*
* `--verbose` can now be specified with the **VERBOSE** macro argument.

For any question about this fork feel free to [contact me](olivier.ldff@gmail.com).

Tested with `qt5.11 qt5.12` on `windows 1803`.

OlivierLDff committed:
fix: build.gradle.in was configure in wrong folder

OlivierLDff committed:
fix ARGS to ARG

Commit:
Use generator expressions for target locations.

Fix issue #26:
- Replace the use of LOCATION properties by generator expressions to acquire dependency target locations.
- Add an extra file(GENERATOR) call to evaluate generator expressions at build time.

Commit:
Add support for the LLVM/Clang toolchain.

nohype committed:
Use generator expressions for the target path.

Instead of the (DEBUG_)LOCATION property, use generator expressions to determine the app's target file path (main shared library).
Fixes issues with generator expressions in the target's source files.

nohype committed:
Use generator expressions for the target path.

Instead of the (DEBUG_)LOCATION property, use generator expressions to determine the app's target file path (main shared library).
Fixes issues with generator expressions in the target's source files.

Commit:
Fixed incorrect tool prefix on Windows

Commit:
Fix QT_ANDROID_APP_PATH begin set too late.

OlivierLDff committed:
fix with new android cmake toolchain that only use ANDROID_STL

OlivierLDff committed:
Merge branch 'master' into pr/1

OlivierLDff committed:
Merge pull request #2 from OlivierLDff/pr/1

Pr/1

OlivierLDff committed:
Update AddQtAndroidApk.cmake

OlivierLDff committed:
mimic build.gradle from qt 5.13

OlivierLDff committed:
fix bracket

OlivierLDff committed:
Dummy CMakeLists.txt that include AddQtAndroidApk.cmake for easier integration with FetchContent_MakeAvailable

OlivierLDff committed:
UPPER_CASE to lower_case

OlivierLDff committed:
Remove tsa args from signing because Android does not care about timestamps in APKs. (source: https://stackoverflow.com/questions/45486415/do-apksigner-supports-timestamp-server-argument )

OlivierLDff committed:
save qt android root path in cache for later use

OlivierLDff committed:
Force variable in cache

OlivierLDff committed:
fix set variable in cache

OlivierLDff committed:
Save QT_ANDROID_SDK_ROOT in cache

Kicer86 committed:
removing entries causing build problems

OlivierLDff committed:
Merge pull request #3 from Kicer86/master

removing entries causing build problems

LaurentGomila commented:
Hi

Sorry that I made you wait for so long 😁 (I hope you're still around).

First, thank you very much for your work. It seems like you addressed many concerns, many of them being necessary for this project to catch up with recent NDK/DSK versions -- and simply continue to work.

However there's really a lot of things here, which is part of why it took me so long to review your PR. Here is what I suggest, in order to move on:
- I will (right after I complete this message) post a review, not a very complete one but hopefully enough to validate the most important points, so we can work on improving the PR (or fix my incorrect understanding of it)
- I'd really like to separate bug fixes from new features, so it's much easier for me to merge fixes and make the project work again, and it's not blocked by concerns about new features, that we may discuss deeper
- *if possible*, all these commits should be squashed into fewer, so that the final commits that get merged into master are "clean" and don't contain your intermediate fixes and additions, just the final result

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -44,8 +45,12 @@ if(NOT QT_ANDROID_NDK_ROOT)
+set(QT_ANDROID_NDK_ROOT ${QT_ANDROID_NDK_ROOT} CACHE STRING "" FORCE)
 message(STATUS "Found Android NDK: ${QT_ANDROID_NDK_ROOT}")
 
+set(QT_ANDROID_CMAKE_FOUND ON CACHE BOOL "QtAndroidCMake have been found" FORCE)
```
What's the reason for adding this variable? And why in cache (i.e. I can switch it OFF)?

OlivierLDff replied:
This variable doesn't have any reason to exist anymore. I used it before discovering usage of FetchContent in CMake.
This variable was helping me to mimic the behavior of FetchContent. In my CMake script i could check if QT_ANDROID_CMAKE_VERSION wasn't defined. If not then download the repository from github. I need to remove this variable.
You can see an example of my previous usage here: https://github.com/OlivierLDff/QaterialGallery/blob/d2f3187c0a15a421b435a7ad15efd75fa9687c5c/platform/PreBuildPlatform.cmake
Yes it's very ugly.

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -44,8 +45,12 @@ if(NOT QT_ANDROID_NDK_ROOT)
 message(STATUS "Found Android NDK: ${QT_ANDROID_NDK_ROOT}")
 
+set(QT_ANDROID_CMAKE_FOUND ON CACHE BOOL "QtAndroidCMake have been found" FORCE)
+set(QT_ANDROID_CMAKE_VERSION "19.0.1" CACHE STRING "QtAndroidCMake version" FORCE)
```
What is that version? Where does it come from? Where is it used?

OlivierLDff replied:
I started to make git tag on some commits.
https://github.com/OlivierLDff/QtAndroidCMake/releases
I started v1 and then i tought it might be a good idea to match ndk version currently in usage. Seeing that now doesn't look like a good idea.
With that variable you could then print the version of the script included into your cmake logs. 
I don't think it's really usefull and the script shouldn't have data that can be found somewhere else. And i mean you should already be aware of the version you are using.

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -1,8 +1,7 @@
 
 # store the current source directory for future use
-set(QT_ANDROID_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
+set(QT_ANDROID_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR} CACHE STRING "Source directory of AddQtAndroidApk.cmake")
```
I'm not a huge fan of adding this (and the others) paths to the cache, it means that they can be edited, which is unnecessary if the user configured its build properly (as described in readme).

Moreover, it adds another set of variables which are more or less aliases to already defined variables, which
in my opinion adds confusion.

And to finish, some of these variables are really meant for internal use only (this one, for example), I don't really see any benefit of exposing them like this.

OlivierLDff replied:
I will try to answer to each of your review.
Sorry for some obscure commits.
Main reason for cache variables, the CMakeLists.txt file adding the script etc... is to be able to use CMake's [FetchContent](https://cmake.org/cmake/help/v3.11/module/FetchContent.html).
Like that in my code i can simply
```
FetchContent_Declare(
    QtAndroidCMake
    GIT_REPOSITORY https://github.com/OlivierLDff/QtAndroidCMake
    GIT_TAG        master
)
FetchContent_MakeAvailable(QtAndroidCMake)
```
The FetchContent_MakeAvailable call "out of the box" the CMakeLists.txt file if any is present. So in my script i have one line to declare the dependency and one to download it. I don't need know exactly the name of the file `*.cmake`. This may seem like a simple answer but it helps when reading my script. Each script i might need is "inserted" into the project the same way.
An example that might work out of the box : https://github.com/OlivierLDff/QaterialGallery/blob/master/platform/PostBuildPlatform.cmake
You can see in there that i created other script QtWindowsCMake and QtIosCMake with the same philosophy.
So long story short, using that `MakeAvailable` function, the variable defined before the macro needs to be stored somewhere when the macro call happen.
Maybe there would be a way to scope them to the file? I added them in the cache as a quick workaround to be able to use MakeAvailable.

LaurentGomila replied:
I see.

That's still confusing to see them and be able to edit them, but if that's the only way then I'll be fine with that.

OlivierLDff replied:
It was a quick fix so maybe there is a better way to handle that

OlivierLDff replied:
A good idea to fix this issue would be to go back the way you were declaring the variable. And store them in cache with same named variables in the CMakeLists.txt. This solution will result in:
* If user simply include the script then no variable get stored in cache
* If user use FetchContent to include the script then variables are going to be stored in cache.
Would you consider this as a "fix"?

LaurentGomila replied:
That would not be perfect, but definitely better, yes :+1:

We should also explicitly mention in the readme file that CMakeLists.txt is meant for this specific use case of FetchContent; for regular use, one should still include AddQtAndroidApk.cmake.

OlivierLDff replied:
I did it in 372ebda76952b30d29c4f1c63f31d2d559637c8e
Where do you think i should mention the use case in the readme?
Just before the How to run CMake header?
Alternative: How to integrate using FetchContent
Something like that?

LaurentGomila replied:
Hmm maybe we should not add anything to the readme, and add a comment about FetchContent directly inside the CMakeLists.txt file instead. Using FetchContent works out of the box with this file, right? If so, we don't need to add anything to the readme file about FetchContent.

Regarding the readme: you should also remove the first paragraph (*"this fork..."*), and put any relevant addition into the dedicated sections.

I would also remove the VERBOSE option and force it by default, like before. Again, let's aim at simplicity.

OlivierLDff replied:
I added the verbose option because it's an option that i exposed in lots of CMake scripts. CMake script are in general non verbose.
I documented the CMakeLists in a2f5f33d390bd4affa2a8a69b3f77a5e0fd5f2e8
Update the Readme in 3eaa6256d0f486a1c637322031287f30fe638166

LaurentGomila commented on CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+cmake_minimum_required(VERSION 3.0.0 FATAL_ERROR)
```
What's the reason for adding a CMakeLists.txt file? Files that provide functions are usually included directly (see standard stuff that CMake provides). CMakeLists.txt files are supposed to add targets to the current project.

OlivierLDff replied:
Purpose of CMakeLists.txt is too include the script when using the FetchContent Declare/MakeAvailable.
Example : https://github.com/OlivierLDff/QaterialGallery/blob/master/platform/PostBuildPlatform.cmake

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+        set(QT_ANDROID_STL_PATH "${QT_ANDROID_NDK_ROOT}/sources/cxx-stl/${ANDROID_STL_PREFIX}/libs/${ANDROID_ABI}/libc++_shared.so")
     else()
         set(QT_ANDROID_STL_PATH)
+        if(ANDROID_STL_STATIC_LIBRARIES)
```
We should remain consistent with the rest of the file, and not use this old CMake syntax (`if(XXX) else(XXX) endif(XXX)`) 😉

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+        set(QT_ANDROID_APP_VERSION_CODE 1)
+    endif()
+
+    if(ARG_VERSION_NAME)
```
We should remain consistent with the rest of the file, and not use this old CMake syntax (`if(XXX) else(XXX) endif(XXX)`) 😉

OlivierLDff replied:
Totally agree i only knew the old cmake way when i did that.

OlivierLDff replied:
Fixed in 73c857cfb89782d77744d5a4d5de10bfee0044f6

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
         if(NOT QT_ANDROID_APP_VERSION)
             set(QT_ANDROID_APP_VERSION ${QT_ANDROID_APP_VERSION_CODE})
         endif()
+    endif(ARG_VERSION_NAME)
```
About this new `VERSION_NAME` option: why would the Android APK have a different version than the project itself? do you have a relevant use case for this addition?

OlivierLDff replied:
I've added this option because at the time i didn't use the `VERSION` property of project. And i think it's nice to expose, for example if you have your internal version used in your project. And a version numbering different for the customer.
Or if you want to expose only a part of the version to your customer. Anyway this field is mostly decorative i think(i might be wrong). But from what i understand, it's the VERSION_CODE that is important to android.

LaurentGomila replied:
If this new option is not really necessary, I prefer to keep things simple. Adding another `VERSION_XXX` variable on top of `VERSION_CODE` and CMake's `PROJECT_VERSION` might be a little confusing. And to me this variable should not be different from the project's version, that's exactly the same thing.

This project is just maintained for my own needs, so I like to keep it "minimal", so that I'm still able to work on it when I need to, even after 2 or 3 years without touching it 😊

OlivierLDff replied:
Yes i see your point, I should start using PROJECT_VERSION and not doing any workaround.

OlivierLDff replied:
This variable is a string so it can be used to decorate the version in Android. For example doing something like "v1.0 beta" or "v1.0 rc1". I don't think you can do such thing with `PROJECT_VERSION`?

LaurentGomila replied:
I don't know, but please don't spend so much time searching artificial use cases. Let's stick to what we need right now, and don't worry, if someone wants to support "beta" or "rc" in its Android version string, I'm sure he'll find his way to this github repo 😉 

I want something simple that works well, not a super-flexible tool that can fit every single use case.

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+
+    if(ARGS_EXTRA_QML)
+        set(QT_ANDROID_QML_IMPORT_PATH ${ARGS_EXTRA_QML})
+    endif()
```
About this new `EXTRA_QML` option: the previous version would scan the whole project tree for QML files; this addition allows to define an explicit path, but is it really necessary? use case?

OlivierLDff replied:
I don't have a use case yet, because i'm using the qtquickcompiler. So all my qmls file are included into the QRC.

LaurentGomila replied:
Great, so let's remove that option.

OlivierLDff replied:
But i'm not sure CMAKE_SOURCE_DIR is also a great default value. Why would it be?
I think main point of that variable is to add Qml modules not in Qt Sdk? 
In your example projects, qml files are included using the .qrc method. In my project i'm using qt quick compiler (so at the end qml files are also in the qrc).
If we remove this variable what would be the way to add an external Qml module?

I'm not a fan of Qml module in general, I prefer to have everything statically linked and embedded into Qrc. But what if someone is using that.

LaurentGomila replied:
> I think main point of that variable is to add Qml modules not in Qt Sdk?

Nope. The `qml-root-path` parameter of `androiddeployqt` is used to scan the project's QML files, in order to detect what plugins and modules they depend on, and copy those libraries to the final APK. So you have to give it a proper value, even if all your QML files are included in the app's qrc. 

`CMAKE_SOURCE_DIR` seems like a good choice because your QML files will most likely be under that root, I can't imagine why one would store its QML files outside the project tree.

OlivierLDff replied:
In my project qml-root-path is set to "" and it still managed to deploy the correct dependencies. (All qml files in the Qt sdk)

LaurentGomila replied:
In mine it didn't work, only some dependencies were found. I had to provide `${CMAKE_SOURCE_DIR}` to make it work. But my project contains two targets: a library and an executable; almost all the QML components are in the library.

OlivierLDff replied:
In my opinion this is the use case of `EXTRA_QML`. To put your folder where you have some extra qml files.
Or you should use the qtquickcompiler from cmake to precompile your file.
In a test project if i put `CMAKE_SOURCE_DIR` as value in the qtdeploy.json file. I end up with bunch of warning because :
* C++ defined modules are not found.
* Qml modules defined in static library are not found
* Private Qml module are not found.

The apk is 600 ko bigger because of useless deployed qml or stuff.
During that test i had in my project directory a `build/` directory that contain the cmake build folder of the app for windows. This folder is under `CMAKE_SOURCE_DIR` so qtandroiddeploy deploy in the apk windows platform related files. Not very clean. 
Ok this is something that happen because i'm in a developement environment . But i don't think the way you set up your Qml dependencies, or the way folder are in your project should impact this tool behavior.
But i also see your point of having a default behavior that work even if not optimized and with warning. But if you want to keep that as a default behavior, we should at least put an option to optimize/remove the warning and useless deployment.

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -97,84 +98,162 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+        else(ANDROID_STL_STATIC_LIBRARIES)
+            message(WARNING "ANDROID_STL_SHARED_LIBRARIES isn't defined, you might need to define ANDROID_STL (${ANDROID_STL}) to a shared stl library (for example c++_shared or gnustl_shared)")
+        endif(ANDROID_STL_STATIC_LIBRARIES)
     endif()
```
This STL part is not clear to me. With default settings (and NDK version 18) I get a warning:

>CMake Warning at dashboard-app/3rdparty/qt-android-cmake/AddQtAndroidApk.cmake:188 (message):
  ANDROID_STL_SHARED_LIBRARIES isn't defined, you might need to define
  ANDROID_STL (c++_static) to a shared stl library and not a static library
Call Stack (most recent call first):
  dashboard-app/CMakeLists.txt:52 (add_qt_android_apk)

Why is `ANDROID_STL_SHARED_LIBRARIES` undefined? Why do I have to define `ANDROID_STL` to a shared STL, and what happens if I don't?

Can't this thing get an implementation that "just works" by default? Or maybe I just need to upgrade my NDK? 😃

OlivierLDff replied:
I don't really remember about this, but i remember it was a mess when i updated to NDK 19.
For example my cmake command look like this:
```
cmake -DCMAKE_PREFIX_PATH=C:/Qt/5.12.1/android_armv7 \
-G "Unix Makefiles" \
-DANDROID_ABI=armeabi-v7a \
-DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake \
-DANDROID_NATIVE_API_LEVEL=27 \
-DANDROID_STL=c++_shared \
-DANDROID_TOOLCHAIN=clang
```
So i as i remember from NDK 19 you need to use `c++_shared`. And if i remember correctly with NDK 18 it was `gnustl_shared`. But i could be totally wrong here.

LaurentGomila replied:
> So i as i remember from NDK 19 you need to use c++_shared. And if i remember correctly with NDK 18 it was gnustl_shared

What I don't understand is, why aren't these the defaults? Is it Qt that forces to use a shared STL? If so, would it be a good idea to set a proper default in the CMake script, according to what you say?

OlivierLDff replied:
So i tried to see if gnustl_shared was still working and it seems not:
```
CMake Error at C:/Users/olivi/AppData/Local/Android/Sdk/android-ndk-r20b/build/cmake/android.toolchain.cmake:249 (message):
  gnustl_shared is no longer supported.  Please switch to either c++_shared
  or c++_static.  See
  https://developer.android.com/ndk/guides/cpp-support.html for more
  information.
```

OlivierLDff replied:
I tried `c++_static` and i get:
```
Invalid STD C++ library name.
```
When it tries to build to build the APK

OlivierLDff replied:
Ok so with NDK20 on my machine with the toolchain i have ANDROID_STL by default set to c++_static. i fixed the script to generate `"stdcpp-path": "C:/Users/olivi/AppData/Local/Android/Sdk/android-ndk-r20b/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a/libc++_static.a",` inside qtdeploy.json.
But it seem that qtandroiddeploy doesn't care about that (at least the .a suffix). I end up in my apk with a file libc++_static.so that is a copy of the correct libc++_static.a.
When i try so launch the app i get the error
```
com.oliv.qaterial E/AndroidRuntime: FATAL EXCEPTION: qtMainLoopThread
    Process: com.oliv.qaterial, PID: 27050
    java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app/com.oliv.qaterial-0EE82Hk--0TzcAuFs5v6Yw==/lib/arm/libc++_static.so" has bad ELF magic
        at java.lang.Runtime.load0(Runtime.java:928)
        at java.lang.System.load(System.java:1633)
        at org.qtproject.qt5.android.QtNative$2.run(QtNative.java:199)
        at org.qtproject.qt5.android.QtThread$2.run(QtThread.java:87)
        at org.qtproject.qt5.android.QtThread$1.run(QtThread.java:61)
        at java.lang.Thread.run(Thread.java:764)
```
It seem that qt for android is compiled by default with c++_shared stl. Problem is that ANDROID_STL is default to c++_static in the NDK toolchain file.
I don't see any other fix than continuing specifying ANDROID_STL=c++_shared when doing the first cmake command.

I will add a warning that say that ANDROID_STL must be set to c++_shared for now.

OlivierLDff replied:
Ok so i fixed it in 1adbdc5575dc6780285de6833da0a071e415e272
Qt is build with c++_shared stl and is expecting it. Like qtandroiddeploy. So even if ANDROID_STL isn't set to c++_shared, libc++_shared.so will be deployed inside the apk with a warning for the dev. 
This will work just fine if ANDROID_STL is the default value c++_static. In that scenario, YourApp.so will be build using c++_static and qt*.so libraries are going to find the required shared stl.

LaurentGomila replied:
Thanks.

> `"You should consider setting ANDROID_STL to c++_shared (like qt)."`

It's Qt, not qt 😆

> "Note that gnustl_shared is no longer supported"

Isn't it redundant with the warning from the toolchain file?

OlivierLDff replied:
You're right.
Fixed in 7ab6d6cfc73764fd6e21d663116d3b8cfa3a7530

LaurentGomila commented on AddQtAndroidApk.cmake:
```diff
@@ -190,15 +269,53 @@ macro(add_qt_android_apk TARGET SOURCE_TARGET)
+            )
+    endif() # QT_ANDROID_MIX_SOURCE_DEPLOY
+
+    if(${CMAKE_BUILD_TYPE} STREQUAL "Release" OR
```
This triggers an error when `CMAKE_BUILD_TYPE` is empty (which is the default), since it makes the IF statement ill-formed. We should either force a suitable default value, or (better) check if it's empty before using it.

OlivierLDff replied:
Yes you are right, i never tried with empty CMAKE_BUILD_TYPE. I think we should just check if CMAKE_BUILD_TYPE is empty and display the warning that is already there. Like that qtandroiddeploy will choose by it's own like it was before. But user will get notified.
What are default qt library used when CMAKE_BUILD_TYPE is empty?

LaurentGomila replied:
> What are default qt library used when CMAKE_BUILD_TYPE is empty?

Seems like it is the release build.

OlivierLDff replied:
Fixed in d70068dabe4d1f97ecae4a8ea84ee30b945ebd45

LaurentGomila reviewed (commented):
In addition to my concerns (which don't prevent the tool to work), my build fails with the following error:

> error: 'orientation|uiMode|screenLayout|screenSize|smallestScreenSize|layoutDirection|locale|fontScale|keyboard|keyboardHidden|navigation' is incompatible with attribute configChanges (attr) flags [locale=4, mcc=1, mnc=2, touchscreen=8, keyboard=16, keyboardHidden=32, navigation=64, orientation=128, screenLayout=256, uiMode=512, screenSize=1024, smallestScreenSize=2048, fontScale=1073741824].
  error: failed processing manifest.

... but this may be a bad SDK/NDK version match (I'm using "old" versions), I'll try to upgrade later.

LaurentGomila commented:
Regarding the build error, I updated my SDK, NDK and provided a suitable `ANDROID_PLATFORM` value in CMake, and the build succeeded :+1:

OlivierLDff committed:
Set QT_ANDROID_BUILD_TYPE to --release by default if no CMAKE_BUILD_TYPE is specified. If CMAKE_BUILD_TYPE wasn't defined the script was breaking because "${CMAKE_BUILD_TYPE} STREQUAL "Release"" was transformed into STREQUAL "Release" for example

OlivierLDff committed:
Warning to indicate that gnustl_shared isn't supported anymore

OlivierLDff committed:
Default QT_ANDROID_STL_PATH to libc++_shared.so to work out of the box from ndk 19 with qt. Even if ANDROID_STL is c++_static for your app.

OlivierLDff committed:
Remove useless echo

OlivierLDff committed:
Remove redundant warning with the android toolchain file from NDK r19 about stl gnustl_shared is no longer supported

OlivierLDff committed:
Remove useless variables QT_ANDROID_CMAKE_FOUND and QT_ANDROID_CMAKE_VERSION

OlivierLDff committed:
QT_ANDROID_SOURCE_DIR QT_ANDROID_QT_ROOT QT_ANDROID_SDK_ROOT QT_ANDROID_NDK_ROOT are kept in cache in CMakeLists.txt that include AddQtAndroidApk.cmake. The variable are longer added in cache inside AddQtAndroidApk.cmake.
By doing that the cache of user only using AddQtAndroidApk.cmake won't have a cache with internal variable.
When including the script by the CMakeLists cache will be populated by internal variable. Maybe there is a better workaround.

OlivierLDff committed:
Match new CMake convention with if/else/endif syntax

OlivierLDff committed:
Doc in CMakeLists.txt on how to use it with FetchContent

OlivierLDff committed:
Update readme to remove information about the fork

LaurentGomila commented:
What's the current status? Is it ok on your side, or are there points that still need to be modified or discussed?

OlivierLDff committed:
Default VERSION_NAME to PROJECT_VERSION if no target version found.

OlivierLDff commented:
Ok so i added some information about the VERSION name in 5f7178eadd7130238f00a9ed4813c5566cbf6f79.

Literal name of the version. If none is provided then the version will be extracted from `my_app` `VERSION` property. If none found then it will be defaulted to `PROJECT_VERSION`. If none is found then it is set to `VERSION_CODE`.

Example:

```cmake
add_qt_android_apk(my_app_apk my_app
    VERSION_NAME "1.2.3"
)
```

So it's defaulted to the behavior you had. You don't need to set it if you don't want to use it.

But the biggest thing we need to solve is the use case of that `EXTRA_QML` property. Defaulting it to CMAKE_SOURCE_DIR add lots of useless deployment in some of my apps and can lead to hard to track errors.

LaurentGomila commented:
I think we should already merge the rest (mainly the improvements and bug fixes), and take some time to discuss new features one by one. I'm really uncomfortable with this PR containing so many different things 😁

This single PR contains many more commits than the full history of the repo, so it would be really cool if we could make a new branch with the result as *a single clean commit*. But without new features that still require discussion or improvements -- let's open new PR(s) for those once the rest is merged. I had a quick look, and it doesn't look like it will require much effort. I can do it if you want, but that won't give you credit for these changes, since I'll be the author of the commit (unless you don't care).

OlivierLDff commented:
Yes sorry for the number of commit, it was my dev branch i did most of that for my personnal use, i tought i will do a PR at a moment.
I never did a pull request with that much stuff. How should i proceed to create single commit?
Should i just refork from your branch and copy and paste code to create commit from that?
Or is there a smart way to "extract" the commit?

LaurentGomila commented:
> Or is there a smart way to "extract" the commit?

You can squash commits and force push to the same branch, to overwrite the whole commit history of the PR, but here you must also remove some stuff, so I'd just make a new branch and copy & paste relevant modifications. This is also a good way to make a final review of your modifications, clean minor stuff, etc. 😃

By the way, I see that you moved the *"Found XXX: [path]"* status messages to CMakeLists.txt; it would be great if they could stay where they were originally (they are quite useful, even for those who don't use FetchContent) 😛

OlivierLDff commented:
> By the way, I see that you moved the "Found XXX: " status messages to CMakeLists.txt; it would be great if they could stay where they were originally (they are quite useful, even for those who don't use FetchContent) 😛
Ok sorry i forgot it was originally there.

Ok i see, another question should i make one branch per PR ?

LaurentGomila commented:
> should i make one branch per PR ?

It doesn't matter how you manage your own repository, as long as it ends as expected here 😉

OlivierLDff commented:
It was more a question what is the best practice to do something like that? I never did it before, and the one branch per PR seems to be the easiest thing to manage.
I will do it tonight.

LaurentGomila commented:
What's your progress?

I need this to be merged for a project at work, that's why I put some pressure on this PR now 😀. As I said, if you don't have the time, I can help.

OlivierLDff commented:
Yes i need some help, i have a lot of work at work and not so much free time. So it's best that you do it your way.

LaurentGomila commented:
No problem :+1:

LaurentGomila commented:
See #32

LaurentGomila commented:
I merged #32 so I'm closing this PR. Don't hesitate to open new PRs or issues if needed.

And thanks again for your work!
